demos/stackcheck/README.md

```diff
@@ -0,0 +1,45 @@
+# Stack Check
+
+In this demo we demonstrate stack checking.
+
+Overflow occurs if the stack pointer exceeds the low address bound of the stack. The most common case is deep recursion. Since we only have limited space for the stack, we may clog the stack when we have too many function calls.
+
+In contrast, underflow is the situation that the stack pointer goes beyond the high address bound of the stack. We might encounter underflow when we return the first and the only function or the number of bytes we intend to write exceeds the high bound of the stack.
+
+It is hard to detect illegal use of memory and it may lead to the corruption of the kernel structure. So each time a thread yields, we do checks to ensure the system can work properly.
+
+To illustrate the concept, we need to introduce the following:
+* `check_stack` (in `src/kernel/thread.c`)
+* `choose_next_thread` (in `src/kernel/thread.c`)
+
+## `check_stack`
+
+This function is used to detect if the thread's stack has been overflowed or underflowed. 
+
+In AMT, we place 2 variables `bottom_canary` and `top_canary` on both sides of the stack in the thread structure. These two variables are initialized to `STACK_CANARY`. If their values have been changed, we know that overflow/underflow has occurred.
+
+If the config `KCFG_DESTROY_ON_STACK_ERR` is enabled we invalid the current thread and load the next thread.
+
+If `KCFG_DESTROY_ON_STACK_ERR` is disabled, we just exit the kernel completely. (marking it as an unexpected exit)
+
+## `choose_next_thread`
```

Comment:
I don't think we should be describing choose_next_thread here.

I was reading it again and found the whole process could be simplified a lot so see https://github.com/DavidSpickett/ARMMultiTasking/commit/8cececfa3b9352f02c4cd7db3f82a23b93a01486 and https://github.com/DavidSpickett/ARMMultiTasking/commit/6e02b7d1f09ce6a9dee8d5d0a22c452f0c348f72 for the new state.

Basically:
* yield happens
* we do check_stack, which might invalidate the thread
* if we're doing a YIELD_ANY, just carry on. We'll never return to current if it had a bad stack
* if we're doing a YIELD_TO we have to be careful not to return to current if it has a bad stack and the yield target isn't valid

Also I removed the call into assembly load_next_thread, it was just confusing things.

Based on that new behavior I would think that k_do_yield is a candidate to be described instead of choose_next_thread. Let me know what you think, assuming the new behavior makes sense.